src/graph/graph.c

```diff
@@ -42,17 +42,39 @@ void _edge_accum(void *_z, const void *_x, const void *_y) {
 }
 
 bool _select_op_free_edge(GrB_Index i, GrB_Index j, GrB_Index nrows, GrB_Index ncols, const void *x, const void *k) {
-    const Graph *g = (const Graph*)k;
+    // K is a uint64_t pointer which points to the address of our graph.
+    const Graph *g = (const Graph *) *((uint64_t*)k);
     const EdgeID *id = (const EdgeID*)x;
     if((SINGLE_EDGE(*id))) {
         DataBlock_DeleteItem(g->edges, SINGLE_EDGE_ID(*id));
     } else {
+        /* Due to GraphBLAS V3.0+ parallelism
+         * _select_op_free_edge will be called twice
+         * Tim: "In my draft parallel GraphBLAS,
+         * most of my codes take 2 passes over the data.
+         * It's what Intel calls the Inspector / Executor model of computing.
+         * Both passes are fully parallel.
+         * The first pass is purely symbolic,
+         * and it figures out where all the output data needs to go.
+         * The 2nd pass fills in the data in the output."
+         *
+         * To avoid double freeing we'll place a marker on the first pass:
+         * ids[0] = INVALID_ENTITY_ID
+         * to be picked up by the second pass, on which the array will be freed. */
```

Comment:
I'm not familiar with the Inspector/Executor model, but it sounds like the second pass should be responsible for all of the work of freeing. (Though it may not matter.)
If we have a single edge, then won't the `if` block at 49 execute twice? Is that safe?

Reply:
it will execute twice but it is safe, `DataBlock_DeleteItem` knows how to handle double free.